File: LIS/lis_jonnung.py
"""
yoonnoon님의 `LISv2`로 해결(외워버림)
"""
def lis_solution(sequence):
    length = len(sequence)
    memo = [1 for _ in range(length)]
    for i in range(1, length):
        for j in range(i):
            if sequence[i] > sequence[j]:
                memo[i] = max(memo[i], memo[j] + 1)

    return max(memo)


if __name__ == '__main__':
    import sys

    rl = lambda: sys.stdin.readline()

    c = int(rl())

    for _ in range(c):
        n = int(rl())
        v = list(map(int, rl().split()))
        print(lis_solution(v))

File: LIS/lis_jonnung.go
package	main

import (
	"os"
	"fmt"
	"bufio"
	"strconv"
	"strings"
)

func max(arr []int64) int64 {
	max := int64(0)
	for _, v := range arr {
		if v > max {
			max = v
		}
	}

	return max
}


func solution(size int64, seqc []int64) {
	var mem [500]int64

	for m := int64(0); m < size; m++ {
		mem[m] = 1
	}

	for i := int64(1); i < size; i++ {
		for j := int64(0); j < i; j++ {
			if seqc[i] > seqc[j] {
				compare := [2]int64{mem[i], mem[j] + 1}
				mem[i] = max(compare[:])
			}
		}
	}

	fmt.Println(max(mem[:size]))
}


func main() {
	var c, n int64
	var v string
	var sequence []int64

	input := bufio.NewScanner(os.Stdin)

	// TC 개수 입력
	input.Scan()
	c, _ = strconv.ParseInt(input.Text(), 0, 64)

	for i := int64(0); i < c; i++ {
		// 원소 개수 입력
		input.Scan()
		n, _ = strconv.ParseInt(input.Text(), 0, 64)

		// 길이가 n인 수열 입력
		input.Scan()
		v = input.Text()
		values := strings.Fields(v)
		sequence = make([]int64, len(values))
		for j, num := range values {
			intNum, _ := strconv.Atoi(num)
			sequence[j] = int64(intNum)
		}

		solution(n, sequence)
	}
}